### PR TITLE
Improve apache-commons-configuration

### PR DESCRIPTION
**Pull request by CheeseHunter117**
I tried improving the existing fuzz targets that were written by me, based on the data gathered by running OSS-Fuzz.
I tried disabling all logging. It seems to work on the majority of Configuration classes.

### CODE REVIEW

File: projects/apache-commons-configuration/YAMLConfigurationReadFuzzer.java [hunk 1]
```diff
@@ -1,22 +1,65 @@
...
+                } else {
+                    inputStream = new ByteArrayInputStream(data.consumeBytes(Integer.MAX_VALUE));
+                    yamlConfig.read(inputStream);
+                }
+            } else {
+                if (useLoaderOptions) {
+                    final LoaderOptions loaderOptions = createLoaderOptions(data);
+                    inputStream = new ByteArrayInputStream(data.consumeBytes(Integer.MAX_VALUE));
+                    reader = new InputStreamReader(inputStream);
+                    yamlConfig.read(reader, loaderOptions);
```
Comment (onionpsy) [inline projects/apache-commons-configuration/YAMLConfigurationReadFuzzer.java:40]: I don't think all this `InputStreamReader` part is needed as `read()` will take any InputStream: https://github.com/apache/commons-configuration/blob/master/src/main/java/org/apache/commons/configuration2/YAMLConfiguration.java#L110

File: projects/apache-commons-configuration/YAMLConfigurationReadFuzzer.java [hunk 2]
```diff
@@ -1,22 +1,65 @@
...
+                } else {
+                    inputStream = new ByteArrayInputStream(data.consumeBytes(Integer.MAX_VALUE));
+                    reader = new InputStreamReader(inputStream);
+                    yamlConfig.read(reader);
```
Comment (onionpsy) [inline projects/apache-commons-configuration/YAMLConfigurationReadFuzzer.java:44]: Same here

File: projects/apache-commons-configuration/YAMLConfigurationReadFuzzer.java [hunk 3]
```diff
@@ -1,22 +1,65 @@
...
+    private static LoaderOptions createLoaderOptions(FuzzedDataProvider data) {
+        LoaderOptions loaderOptions = new LoaderOptions();
+        loaderOptions.setAllowDuplicateKeys(data.consumeBoolean());
+        loaderOptions.setWrappedToRootException(data.consumeBoolean());
+        loaderOptions.setAllowRecursiveKeys(data.consumeBoolean());
+        loaderOptions.setProcessComments(data.consumeBoolean());
+        loaderOptions.setEnumCaseSensitive(data.consumeBoolean());
+
+        // seems to not yet be implemented in the version that apache-commons-configuration2 uses (2022-08-10)
+        // loaderOptions.setNestingDepthLimit(data.consumeInt(50, 100)); // 50 as minimum as that is the default value
```
Comment (onionpsy) [inline projects/apache-commons-configuration/YAMLConfigurationReadFuzzer.java:61]: Please remove unused comment & code

File: projects/apache-commons-configuration/YAMLConfigurationWriteFuzzer.java [hunk 4]
```diff
@@ -1,25 +1,68 @@
...
+            final InputStream inputStream;
+            final InputStreamReader reader;
+            if (useInputStream) {
+                if (useLoaderOptions) {
+                    final LoaderOptions loaderOptions = createLoaderOptions(data);
+                    inputStream = new ByteArrayInputStream(data.consumeBytes(Integer.MAX_VALUE));
+                    yamlConfig.read(inputStream, loaderOptions);
+                } else {
+                    inputStream = new ByteArrayInputStream(data.consumeBytes(Integer.MAX_VALUE));
+                    yamlConfig.read(inputStream);
```
Comment (onionpsy) [inline projects/apache-commons-configuration/YAMLConfigurationWriteFuzzer.java:34]: Same remark as in your `YAMLConfigurationReadFuzzer.java` fuzz target

File: projects/apache-commons-configuration/YAMLConfigurationWriteFuzzer.java [hunk 5]
```diff
@@ -1,25 +1,68 @@
...
+    private static LoaderOptions createLoaderOptions(FuzzedDataProvider data) {
+        LoaderOptions loaderOptions = new LoaderOptions();
+        loaderOptions.setAllowDuplicateKeys(data.consumeBoolean());
+        loaderOptions.setWrappedToRootException(data.consumeBoolean());
+        loaderOptions.setAllowRecursiveKeys(data.consumeBoolean());
+        loaderOptions.setProcessComments(data.consumeBoolean());
+        loaderOptions.setEnumCaseSensitive(data.consumeBoolean());
+
+        // seems to not yet be implemented in the version that apache-commons-configuration2 uses (2022-08-10)
+        // loaderOptions.setNestingDepthLimit(data.consumeInt(50, 100)); // 50 as minimum as that is the default value
```
Comment (onionpsy) [inline projects/apache-commons-configuration/YAMLConfigurationWriteFuzzer.java:64]: Please remove unused comment & code

File: projects/apache-commons-configuration/XMLConfigurationWriteFuzzer.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+        xmlConfig.setLogger(null); // disable the logger
+
+        final FileHandler fileHandler = new FileHandler(xmlConfig);
+        fileHandler.setPath(absoluteFilepath);
+
+        final StringWriter writer = new StringWriter();
+
+        try {
+            fileHandler.load();
+            xmlConfig.write(writer);
```
Comment (onionpsy) [inline projects/apache-commons-configuration/XMLConfigurationWriteFuzzer.java:42]: Shouldn't you also add fuzz data in the `StringWriter`?

**Review by onionpsy — CHANGES_REQUESTED** [review]

File: projects/apache-commons-configuration/YAMLConfigurationReadFuzzer.java (hunk 1 above)
Comment (CheeseHunter117) [reply projects/apache-commons-configuration/YAMLConfigurationReadFuzzer.java:40]: I don't think I understand the problem.
Yes, `read()` can take any `InputStream`, but there is also an overloaded variant takes takes a `Reader` and I wanted to increase the coverage of my fuzz target.
Should I refrain from doing that?

Comment (onionpsy) [reply projects/apache-commons-configuration/YAMLConfigurationReadFuzzer.java:40]: Yes, but it all use the same code. `read(final Reader in)` is the same as `read(final InputStream in)`. Snakeyaml is handling the different types here.

Comment (CheeseHunter117) [reply projects/apache-commons-configuration/YAMLConfigurationReadFuzzer.java:40]: Okay, noted.

File: projects/apache-commons-configuration/XMLConfigurationWriteFuzzer.java (hunk 6 above)
Comment (CheeseHunter117) [reply projects/apache-commons-configuration/XMLConfigurationWriteFuzzer.java:42]: I don't think so.
My idea was to see if the parser could read the data but show bugs while writing it back.
I am not sure if adding data to the writer before or after calling `xmlConfig.write` would change anything.

Comment (onionpsy) [reply projects/apache-commons-configuration/XMLConfigurationWriteFuzzer.java]: Alright

**Comment by CheeseHunter117** [comment]
@onionpsy I have fulfilled all requests.

**Comment by onionpsy** [comment]
One last thing, could you also please add the Google header? Example here https://github.com/google/oss-fuzz/blob/master/projects/json-sanitizer/ValidJsonFuzzer.java, just change the date.

**Comment by CheeseHunter117** [comment]
I am sorry to realise this only now, but I opened this PR to the wrong branch 😓.
Is it okay to create PR to Google with 2 projects inside or should we try fixing this?

**Comment by onionpsy** [comment]
No worries. I also didn't properly verify.

I think it should be fine, just mention it in the PR

**Pull request merged**